File: src/main/scala/com/kotori316/fluidtank/integration/ae2/AEFluidInv.java
package com.kotori316.fluidtank.integration.ae2;

import appeng.api.config.Actionable;
import appeng.api.networking.security.IActionSource;
import appeng.api.stacks.AEFluidKey;
import appeng.api.stacks.AEKey;
import appeng.api.stacks.KeyCounter;
import appeng.api.storage.MEStorage;
import net.minecraft.network.chat.Component;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import scala.Option;
import scala.jdk.javaapi.OptionConverters;

import com.kotori316.fluidtank.fluids.FluidAction;
import com.kotori316.fluidtank.fluids.FluidAmount;
import com.kotori316.fluidtank.fluids.FluidKey;
import com.kotori316.fluidtank.fluids.VariantUtil;
import com.kotori316.fluidtank.tiles.TileTank;

record AEFluidInv(TileTank tank) implements MEStorage {

    @Override
    public long insert(AEKey what, long amount, Actionable actionable, IActionSource source) {
        FluidAmount fluidAmount = fromAEStack(what, amount);
        FluidAmount filled = tank.connection().handler().fill(fluidAmount, actionable == Actionable.MODULATE ? FluidAction.EXECUTE : FluidAction.SIMULATE);
        return filled.fabricAmount();
    }

    @Override
    public long extract(AEKey what, long amount, Actionable actionable, IActionSource source) {
        FluidAmount fluidAmount = fromAEStack(what, amount);
        FluidAmount drained = tank.connection().handler().drain(fluidAmount, actionable == Actionable.MODULATE ? FluidAction.EXECUTE : FluidAction.SIMULATE);
        return drained.fabricAmount();
    }

    @Override
    public void getAvailableStacks(KeyCounter out) {
        var amount = OptionConverters.toJava(tank.connection().getFluidStack());
        amount.ifPresent(fluidAmount -> out.add(AEFluidKey.of(fluidAmount.fluid(), OptionConverters.toJava(fluidAmount.nbt()).orElse(null)),
            Math.min(fluidAmount.fabricAmount(), Long.MAX_VALUE - Integer.MAX_VALUE * 2L)));
    }

    @NotNull
    static FluidAmount fromAEStack(@Nullable AEKey what, long amount) {
        if (what instanceof AEFluidKey fluidKey) {
            return FluidAmount.apply(fluidKey.getFluid(), amount, Option.apply(fluidKey.copyTag()));
        } else {
            return FluidAmount.EMPTY();
        }
    }

    @Override
    public Component getDescription() {
        return this.tank.getName();
    }

    @SuppressWarnings("UnstableApiUsage")
    @Override
    public boolean isPreferredStorageFor(AEKey what, IActionSource source) {
        if (what instanceof AEFluidKey fluidKey) {
            var tankKey = FluidKey.from(tank.connection().fluidType());
            if (tankKey.fluid().isSame(fluidKey.getFluid())) {
                if (tankKey.tag().isEmpty() && !fluidKey.hasTag())
                    return true; // No NBT is in stack.
                else
                    return fluidKey.matches(VariantUtil.convert(tankKey)); // Check NBT as FluidStack
            }
        }
        return false;
    }
}

File: src/main/scala/com/kotori316/fluidtank/integration/ae2/AEFluidInv.scala
package com.kotori316.fluidtank.integration.ae2

import appeng.api.config.Actionable
import appeng.api.networking.security.IActionSource
import appeng.api.stacks.{AEFluidKey, AEKey, KeyCounter}
import appeng.api.storage.MEStorage
import com.kotori316.fluidtank.Utils
import com.kotori316.fluidtank.fluids.{FluidAmount, FluidKey}
import com.kotori316.fluidtank.tiles.TileTank
import net.minecraft.network.chat.Component
import net.minecraftforge.fluids.FluidAttributes

case class AEFluidInv(tank: TileTank) extends MEStorage {
  override def insert(what: AEKey, amount: Long, actionable: Actionable, source: IActionSource): Long = {
    val fluidAmount = fromAEStack(what, amount)
    val filled = tank.connection.handler.fill(fluidAmount, actionable.getFluidAction)
    filled.amount
  }

  override def extract(what: AEKey, amount: Long, actionable: Actionable, source: IActionSource): Long = {
    val fluidAmount = fromAEStack(what, amount)
    val drained = tank.connection.handler.drain(fluidAmount, actionable.getFluidAction)
    drained.amount
  }

  override def getAvailableStacks(out: KeyCounter): Unit = {
    val amount = tank.connection.getFluidStack
    amount.foreach(fluidAmount =>
      out.add(AEFluidKey.of(fluidAmount.fluid, fluidAmount.nbt.orNull), Math.min(fluidAmount.amount, Long.MaxValue - Integer.MAX_VALUE * 2L)))
  }

  private def fromAEStack(what: AEKey, amount: Long): FluidAmount =
    what match {
      case fluidKey: AEFluidKey => FluidAmount.fromStack(fluidKey.toStack(Utils.toInt(amount))).setAmount(amount)
      case _ => FluidAmount.EMPTY
    }

  override def getDescription: Component = this.tank.getName

  override def isPreferredStorageFor(what: AEKey, source: IActionSource): Boolean = {
    what match {
      case fluidKey: AEFluidKey =>
        val tankKey = FluidKey.from(tank.connection.fluidType)
        if (tankKey.fluid.isSame(fluidKey.getFluid))
          tankKey.tag.isEmpty && !fluidKey.hasTag // No NBT is in stack.
        else
          fluidKey.matches(tankKey.createStack(FluidAttributes.BUCKET_VOLUME)) // Check NBT as FluidStack
      case _ => false
    }
  }
}
